es6-shim.js

```diff
@@ -665,7 +666,7 @@
   var addIterator = function (prototype, impl) {
     var implementation = impl || function iterator() { return this; };
     defineProperty(prototype, $iterator$, implementation);
-    if (!prototype[$iterator$] && Type.symbol($iterator$)) {
+    if (prototype && !prototype[$iterator$] && Type.symbol($iterator$)) {
```

Comment:
similarly, this change is fine but when is `addIterator` called with a falsy `prototype` value??

Reply:
When the `object` variable in the `defineProperty` method is marked as `undefined`, the `prototype` is also marked as `undefined`.

Reply:
Why would addIterator ever get called in this case tho?

Reply:
I captured `arguments.callee.caller` inside `addIterator` method to check this.
Most of the `addIterator` calls were made from top-level functions. The reason why `addIterator` is called seems to be because there are no more errors in `defineProperty`.

Reply:
I don't think there's anything strange about `addIterator` being called.

Reply:
It is absolutely a bug in es6-shim if `addIterator` is called with non-object arguments. That should be fixed directly - this approach is just papering over the real problem.

Reply:
I tracked the problem and found the following results.

```
StringIterator.prototype: object
String.prototype: object
ArrayIterator.prototype: object
Array.prototype: object
Object.getPrototypeOf([].values()): undefined
```

Reply:
What does `[].values.toString()` print out?

Reply:
```
* typeof([].values)
function

* [].values.toString()
function values() {
      return new ArrayIterator(this, 'value');
    }
```

Reply:
Hmm, I'm super confused. ArrayIterator is a normal function, and a new instance of it should just inherit from Object.prototype, which should itself inherit from null. What does `Object.getPrototypeOf(Object.prototype)` print out?

Reply:
```
* typeof(Object.getPrototypeOf(Object.prototype))
undefined

* Object.getPrototypeOf(Object.prototype)
[BLANK]
```

Reply:
I think it's a problem because the engine doesn't have the `__proto__` keyword.

```
* typeof([].values().__proto__)
undefined

* ('__proto__' in [].values()).toString()
false
```

Maybe we need a polyfill for `__proto__`. Or I think we need to find an alternative way (because it is deprecated)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto

Reply:
Ahhh - when `__proto__` doesn't exist, `Object.getPrototypeOf` probably shouldn't be shammed at all. However, there's supposed to be a special exception in the sham so that `Object.getPrototypeOf({})` returns `Object.prototype` - but it _doesn't_ have a special exception so that getting the proto of Object.prototype returns `null`. Let me think a bit about what to do here.

Reply:
if the only change made is to the `if` block on line 1270, to something like:
```js
if (Object.getPrototypeOf) {
  var ChromeArrayIterator = Object.getPrototypeOf([].values());
  if (ChromeArrayIterator) { // in WSH, this is `undefined`
    addIterator(ChromeArrayIterator);
  }
}
```

does that fix the issue?

Reply:
Fixed! There are no errors.

Reply:
awesome! let's update this PR to make that change, and we can get this merged

Reply:
Thanks :)